File: src/leetcode/python/BinaryTreeTilt_563.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BinaryTreeTilt_563(object):
    def findTilt(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        res = [0]

        def sufTrav(node):
            if not node:
                return 0
            l = sufTrav(node.left)
            r = sufTrav(node.right)
            res[0] += abs(l - r)
            return node.val + l + r

        if not root:
            return 0
        return abs(sufTrav(root.left) - sufTrav(root.right)) + res[0]

File: src/leetcode/java/BinaryTreeTilt_563.java
package leetcode.java;

public class BinaryTreeTilt_563 {
  public int findTilt(TreeNode root) {
    int[] res = new int[1];
    return Math.abs(sufTrav(root.left, res) - sufTrav(root.right, res)) + res[0];
  }

  private int sufTrav(TreeNode node, int[] res) {
    if (node == null) return 0;
    int l = sufTrav(node.left, res);
    int r = sufTrav(node.right, res);
    res[0] += Math.abs(l + r);
    return node.val + l + r;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }
}
